### PR TITLE
Fix search sidebar a11y issues

### LINKED ISSUE #1168: [Deque Analysis] Search Results "critical" accessibility issues
> Critical: This issue results in blocked content for people with disabilities, and will definitely prevent them from accessing fundamental features or content. This type of issue puts your organization at risk. Prioritize fixing as soon as possible, within the week if possible. Remediation should be a top priority. 

## Deque Analysis Summary

A total of 24 Critical issues exist in the Search Results (https://demo7.dspace.org/search?spc.page=1&query=test) . They are summarized here:
1. (20 issues) All Filter checkboxes are missing labels.  The values next to the checkboxes are improperly in `<span>` tags instead of `<label>` tags which reference the checkbox input field.  See for example: https://www.a11ymatters.com/pattern/checkbox/#associating-the-label-with-the-input
    * Alternatively, they can be kept as `<span>`, but they need an `id` attribute which can be referenced by `aria-labelledby` (e.g. `<input type="checkbox" aria-labelledby="text"><span id="text">Text</span>`
    * Issue IDs: 468524, 468525, 468526, 468527, 468528, 468529, 468530, 468531, 468532, 468533, 468534, 468535, 468536, 468537, 468538, 468539, 468540, 468541, 468542, 468543
2. (ID 468956) All Filter free-text entry fields are missing labels (e.g. "Author name", "Subject", Date fields, etc).  We should either add a `label` above each, or add a `aria-label` attribute to describe them.
     * In a separate "serious" ticket (ID 468933), Deque recommends that these fields have "visible labels" as their purpose is not clear.  Perhaps we should create a small label saying "Search filter values" or similar?
3. (ID 468903) Expandable/Accordion sections (filters) are not keyboard accessible. There's no way to expand them using just keyboard.  Here's an example of how this might be achieved: https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html  (Others may exist as well, this was just the first I found that seemed similar)
4. (ID 468906) "Show More" and "Collapse" links (in filter sidebar) have no `href` value.
    * Fix this issue by providing an HREF value (even if JavaScript event handlers are used) for the link. (Note: Ensure that if JavaScript event handlers are used, they allow the link to be activated with the Enter key in addition to mouse click.)
    * e.g. `<a href=quickview.html">Quick view</a>`
    * e.g. `<a href="javascript:void(0);">Quick view</a>`
5. (ID 469004) Date slider increase/decrease buttons are missing a focus indicator. All focusable elements MUST have a visual focus indicator when in focus.
    * Common keyboard focus indicator solutions include (choose one): (1) A 2px box around the focused element, (2) A change in the background color of the focused element (3) The addition of an icon, such as an arrow, next to a menu item (4) The addition of a thick line under or next to the focused element (5) A distinctive change to the text of the component such as making it bold and/or underlined (6) Use the default browser focus indicator. NOTE: Since each browser has its own default focus indicator style, check each of the major browsers (Chrome, Firefox, Edge, IE, Safari) to ensure the default focus indicator is visible as you tab through the page. 
     * See also https://dequeuniversity.com/class/input-methods2/keyboard-input/visual-focus-indicator

The full list of these issues (which includes links for more info) can be found at this link (requires login): https://axeauditor.dequecloud.com/test-run/0856438a-a19a-11eb-bc31-b7d5be387c86/issues?activeTab=dt-issue&page=0&pageSize=100&sortField=ordinal&sortDir=asc&row=9&filter%5Bseverity%5D=4&filter%5Bpage_number%5D=1

## More Information / Tools

* Consider using [Deque's free Chrome plugin](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) to check your work

### PR DESCRIPTION
**Pull request by ybnd**
## References

* Fixes #1168

## Description

Addressed accessibility issues in the search sidebar

* Added labels
* Added `null` href values to all "empty" links for screen readers
* Improved keyboard navigation
  * Expand/collapse filters with keyboard; highlight facet when selected
  * Exclude controls in collapsed facets from the tablist

## Instructions for Reviewers

**Testing:**

1. Go to `/search`
2. Open some filter facets. 
   * [ ] Search boxes should have labels
   * [ ] Date range boxes should have labels
3. Navigate the sidebar using `Tab`/`Shift+Tab`
   * [ ] You should be able to expand/collapse facets with the keyboard
   * [ ] When a closed facet is selected, pressing `Tab` should select the next facet (old behaviour was to start selecting the hidden controls in the selected facet instead)

## Checklist

_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Comment by ybnd** [comment]
Side note:

Inconsistent keyboard navigation between Firefox & Chrome

* Chrome: Works as exected: all buttons & links can be selected with the keyboard
* Firefox: `<a>` elements are not focusable; multiple links & buttons can not be used with the keyboard
  * List/grid selector
  * More/collapse toggle links
  * `<ds-search-facet-range-option>`
  * Search result elements
  * Page controls

**Comment by tdonohue** [comment]
@ybnd : If you could rebase this on the latest `main`, the CI/build process has just been fixed.  Once you pull in the fix, your tests should hopefully succeed.

File: src/app/shared/search/search-filters/search-filter/search-filter.component.html [hunk 1]
```diff
@@ -1,17 +1,19 @@
-<div class="facet-filter d-block mb-3 p-3" *ngIf="active$ | async">
-  <div (click)="toggle()" class="filter-name">
+<div class="facet-filter d-block mb-3 p-3" *ngIf="active$ | async"
+     [id]="regionId" [attr.aria-labelledby]="toggleId" [ngClass]="{ 'focus': focusBox }" role="region">
+  <button (click)="toggle()" (focusin)="focusBox = true" (focusout)="focusBox = false"
+          class="filter-name d-flex" [attr.aria-controls]="regionId" [id]="toggleId" [attr.aria-expanded]="false">
```
Comment (tdonohue) [inline src/app/shared/search/search-filters/search-filter/search-filter.component.html:4]: I'm seeing a warning in Deque's Axe DevTools (in Chrome) that all of these new `<button>` tags have the same `id` and same `aria-controls` attribute.  When I inspect these buttons in my browser, I see that's correct...somehow they all end up with the same `id` and `aria-controls` values, even though it appears you meant for them to be unique.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@ybnd : Overall, this looks correct to me.  However, I've noticed a few minor (at least I think they are minor) issues when re-scanning the search results page using [Deque's Axe DevTools](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US).   If it's possible to do so, I'd appreciate it if we could fix these two small things before merging.  Thanks!

File: src/app/shared/search/search-filters/search-filter/search-filter.component.html [hunk 2]
```diff
@@ -1,17 +1,19 @@
...
     <h5 class="d-inline-block mb-0">
       {{'search.filters.filter.' + filter.name + '.head'| translate}}
     </h5>
-    <span class="filter-toggle fas float-right"
+    <span class="filter-toggle flex-grow-1 fas p-auto"
           [ngClass]="(collapsed$ | async) ? 'fa-plus' : 'fa-minus'"
           [title]="((collapsed$ | async) ? 'search.filters.filter.expand' : 'search.filters.filter.collapse') | translate"
           [attr.aria-label]="((collapsed$ | async) ? 'search.filters.filter.expand' : 'search.filters.filter.collapse') | translate">
```
[the remark this answers is not in the document]
Comment (ybnd) [reply src/app/shared/search/search-filters/search-filter/search-filter.component.html]: Fixed.

File: src/app/shared/search/search-filters/search-filter/search-filter.component.html (hunk 1 above)
Comment (ybnd) [reply src/app/shared/search/search-filters/search-filter/search-filter.component.html]: I'm not sure how that's happening tbh... 
Thought `this.constructor['ɵcmp'].id` thing in those getters was supposed to be unique for each component instance. @artlowel any idea what's going on?

Comment (artlowel) [reply src/app/shared/search/search-filters/search-filter/search-filter.component.html]: It turns out those IDs are only unique for the _type_ of component, not for the instance. So we'll write a service instead, something akin to a sequence in SQL that lets you retrieve a unique integer every time you call next() on it.

Comment (ybnd) [reply src/app/shared/search/search-filters/search-filter/search-filter.component.html]: ↑ Done.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @ybnd and @artlowel .  Retested today & re-scanned using the Axe DevTools, and all the previous "critical" accessibility issues in the search sidebar appear to be fixed.  Thanks again!

**Pull request merged**